In [1]:
import pandas as pd
import numpy as np


In [2]:
userdf=pd.read_json("/content/users.json", lines=True)

In [3]:
userdf

_id  active               createdDate  \
0    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3    {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
..                                    ...     ...                       ...   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSource state  
0    {'$date': 1609687537858}     consumer        Email    WI  
1    {'$date': 1609687537858}     consumer        Email    WI  
2    {'$date': 1609687537858}     consumer        Email    WI  
3    {'$date': 1609687530597}     consumer        Email    WI  
4    {'$date': 1609687537858}     consumer        Email    WI  
..                        ...          ...          ...   ...  
490  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
491  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
492  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
493  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
494  {'$date': 1614963143204}  fetch-staff          NaN   NaN  

[495 rows x 7 columns]

In [5]:
userdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


In [6]:
userdf.describe()

_id active               createdDate  \
count                                    495    495                       495   
unique                                   212      2                       212   
top     {'$oid': '54943462e4b07e684157a532'}   True  {'$date': 1418998882381}   
freq                                      20    494                        20   

                       lastLogin      role signUpSource state  
count                        433       495          447   439  
unique                       172         2            2     8  
top     {'$date': 1614963143204}  consumer        Email    WI  
freq                          20       413          443   396

In [8]:
missing_values = userdf.isnull().sum()
missing_values

_id              0
active           0
createdDate      0
lastLogin       62
role             0
signUpSource    48
state           56
dtype: int64

In [9]:
missing_values_percent = missing_values * 100 / len(userdf)
print(missing_values_percent)

_id              0.000000
active           0.000000
createdDate      0.000000
lastLogin       12.525253
role             0.000000
signUpSource     9.696970
state           11.313131
dtype: float64


In [10]:
userdf['_id'] = userdf['_id'].apply(str)
userdf['createdDate'] = userdf['createdDate'].apply(str)
userdf['lastLogin'] = userdf['lastLogin'].apply(str)


duplicate_rows = userdf[userdf.duplicated()]
duplicate_rows

_id  active               createdDate  \
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
5    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
8    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
..                                    ...     ...                       ...   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSource state  
1    {'$date': 1609687537858}     consumer        Email    WI  
2    {'$date': 1609687537858}     consumer        Email    WI  
4    {'$date': 1609687537858}     consumer        Email    WI  
5    {'$date': 1609687537858}     consumer        Email    WI  
8    {'$date': 1609687537858}     consumer        Email    WI  
..                        ...          ...          ...   ...  
490  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
491  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
492  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
493  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
494  {'$date': 1614963143204}  fetch-staff          NaN   NaN  

[283 rows x 7 columns]

In [12]:
total_users = len(userdf)

# Calculate the percentage of users per state and print the result
state_percentages = userdf['state'].value_counts().apply(lambda count: round(count / total_users * 100, 2)).reset_index()
state_percentages.columns = ['state', 'percentage']

# Print the results
state_percentages.apply(lambda row: print(f"State: {row['state']}, Percentage: {row['percentage']}%"), axis=1)

State: WI, Percentage: 80.0%
State: NH, Percentage: 4.04%
State: AL, Percentage: 2.42%
State: OH, Percentage: 1.01%
State: IL, Percentage: 0.61%
State: KY, Percentage: 0.2%
State: CO, Percentage: 0.2%
State: SC, Percentage: 0.2%


0    None
1    None
2    None
3    None
4    None
5    None
6    None
7    None
dtype: object

In [15]:
num_unique_user_ids = userdf["_id"].nunique()
print(num_unique_user_ids)

# Print the unique values of the 'signUpSource' column
unique_sign_up_sources = userdf["signUpSource"].drop_duplicates().values
print(unique_sign_up_sources)

# Print the unique values of the 'state' column
unique_states = userdf["state"].drop_duplicates().values
print(unique_states)

# Print the count of each unique value in the 'signUpSource' column
sign_up_source_counts = userdf["signUpSource"].value_counts().to_dict()
print(sign_up_source_counts)

# Print the count of each unique value in the 'state' column
state_counts = userdf["state"].value_counts().to_dict()
print(state_counts)

212
['Email' 'Google' nan]
['WI' 'KY' 'AL' 'CO' 'IL' nan 'OH' 'SC' 'NH']
{'Email': 443, 'Google': 4}
{'WI': 396, 'NH': 20, 'AL': 12, 'OH': 5, 'IL': 3, 'KY': 1, 'CO': 1, 'SC': 1}


Based on the provided data, it appears that the majority of users who signed up are from Wisconsin. We've identified two main data quality issues:

Redundant Records: A significant portion of user records, more than half of the total 495, are duplicates. This redundancy could lead to skewed analysis and ineffective marketing strategies.
Minor Missing Values: There are also a small percentage of missing values in the 'signUpSource' and 'state' columns, although this issue is less urgent compared to the redundancy problem.
These data quality issues should be addressed to ensure accurate analysis and more efficient use of resources in marketing efforts.